In [1]:
import keras

In [3]:
keras.__version__

'2.15.0'

## Model formation

In [4]:
# Sequential: build anti-object
from keras.models import Sequential

In [5]:
model = Sequential()

In [6]:
from keras.layers import Dense # Dense: fully connected layers

# add layers
# units = neurons
# activation = activation function
# input_dim = input vector lengths
model.add(Dense(units = 64, activation = "relu", input_dim = 100))

# output layer
# units = classes of output
# softmax because multi-class classfication
model.add(Dense(units = 5, activation = "softmax"))

In [7]:
# show model architecture
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 64)                6464      
                                                                 
 dense_1 (Dense)             (None, 5)                 325       
                                                                 
Total params: 6789 (26.52 KB)
Trainable params: 6789 (26.52 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


dense = input layers
dense_1 = output layers
param # includes the biases

In [12]:
# different way to construct model
from keras.layers import Activation
model_1 = Sequential()
model_1.add(Dense(units = 64, input_dim = 100))
model_1.add(Activation("relu"))
model_1.add(Dense(units = 5, activation = "softmax"))
model_1.summary()

Model: "sequential_5"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_6 (Dense)             (None, 64)                6464      
                                                                 
 activation (Activation)     (None, 64)                0         
                                                                 
 dense_7 (Dense)             (None, 5)                 325       
                                                                 
Total params: 6789 (26.52 KB)
Trainable params: 6789 (26.52 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [13]:
# different way to construct model
config = [Dense(units = 10, activation = "relu", input_dim = 100),
          Dense(units = 5, activation = "softmax")]
model_2 = Sequential(config)
model_2.summary()

Model: "sequential_6"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_8 (Dense)             (None, 10)                1010      
                                                                 
 dense_9 (Dense)             (None, 5)                 55        
                                                                 
Total params: 1065 (4.16 KB)
Trainable params: 1065 (4.16 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [18]:
def mymodel(n_layers, n_neuron_per_layer, input_dim, output_dim):
    model = Sequential([Dense(units = n_neuron_per_layer, activation = "relu", input_dim = input_dim)])
    for l in range(n_layers):
        model.add(Dense(units = n_neuron_per_layer, activation = "relu"))
    model.add(Dense(units = output_dim, activation = "softmax"))
    return model

In [19]:
model_3 = mymodel(5, 4, 100, 5)

In [20]:
model_3.summary()

Model: "sequential_9"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_14 (Dense)            (None, 4)                 404       
                                                                 
 dense_15 (Dense)            (None, 4)                 20        
                                                                 
 dense_16 (Dense)            (None, 4)                 20        
                                                                 
 dense_17 (Dense)            (None, 4)                 20        
                                                                 
 dense_18 (Dense)            (None, 4)                 20        
                                                                 
 dense_19 (Dense)            (None, 4)                 20        
                                                                 
 dense_20 (Dense)            (None, 5)                

first layer = 404 param
<br>five hidden layers = 20 param
<br>output layer = 25 param

## Optimiser

In [21]:
# loss = loss function used during training
# optimizer = "sgd" = stochastic gradient descent
# metrics = evaluation metrics show in the log
model.compile(loss = "categorical_crossentropy", optimizer = "sgd", metrics = ["accuracy"])